src/dasi/shelf/ShelfCatalogue.cc

```diff
@@ -0,0 +1,74 @@
+
+#include "dasi/shelf/ShelfCatalogue.h"
+
+#include "dasi/api/Config.h"
+
+#include "dasi/core/FileReadHandle.h"
+#include "dasi/core/SplitReferenceKey.h"
+
+#include <cctype>
+#include <fstream>
+#include <iomanip>
+#include <sstream>
+
+namespace dasi::shelf {
+
+static std::string encode(const std::string_view& v) {
+    std::ostringstream r;
+    r.fill('0');
+    r << std::hex;
+
+    for (const auto c : v) {
+        auto cu = static_cast<unsigned char>(c);
+        if (std::isalnum(cu) || c == '-' || c == '.' || c == '_')
+            r << c;
+        else
+            r << ':' << std::setw(2) << int(cu);
+    }
+
+    return r.str();
+}
+
+ShelfCatalogue::ShelfCatalogue(const core::OrderedReferenceKey& dbkey, const api::Config& config) :
+    Catalogue(dbkey, config),
+    root_(config.getString("root")) {}
+
+
+void ShelfCatalogue::archive(const core::SplitReferenceKey& key, const void* data, size_t length) {
+    ASSERT(key[0] == dbkey());
+    auto path = buildPath(key);
+    std::filesystem::create_directories(path.parent_path());
+    std::ofstream fout(path, std::ios::out | std::ios::binary);
+    fout.write(reinterpret_cast<const char*>(data), length);
+}
+
+api::ObjectLocation ShelfCatalogue::retrieve(const core::SplitReferenceKey& key) {
+    ASSERT(key[0] == dbkey());
+    auto path = buildPath(key);
+    if (!std::filesystem::exists(path)) {
+        std::ostringstream oss;
+        oss << key;
+        throw util::ObjectNotFound(oss.str(), Here());
+    }
+    return core::FileHandleBuilder::toLocation(path);
+}
+
+
+std::filesystem::path ShelfCatalogue::buildPath(const core::SplitReferenceKey& key) {
+    std::filesystem::path p(root_);
+    for (int level = 0; level < 3; ++level) {
```

Comment:
For the semantics of DASI, we really at the very least ought to have level zero corresponding to a directory, rather than just being a flat list of files.

Reply:
I'm not sure I understand what you mean. Given a root directory (in the config file), the first two levels define the directory structures and only the last one ends up in the file name

Reply:
I'm not reading carefully enough. I missed the /=